### PR TITLE
[FIX] account,base,web: fix "Send & Print" on invoices

### PR DESCRIPTION
**Pull request by epictete**
The purpose of the task is twofold:

1. Remove empty lines in the company address.

Until now, the address format was fixed, which could
lead to empty lines if one or more field(s) were missing.
We are now removing empty fields to avoid that.

2. Make sure the external report layout is configured
before generating the PDF.

This will ensure that the company data will appear
in the file. If no layout is defined,
it would not be shown.

task-2834517

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_actions_report.py [hunk 1]
```diff
@@ -1002,12 +1002,15 @@ def report_action(self, docids, data=None, config=True):
...
-            action['context'] = py_ctx
-            return action
+            return self._action_configure_external_report_layout(report_action)
 
         return report_action
+
+    def _action_configure_external_report_layout(self, report_action=None):
+        action = self.env["ir.actions.actions"]._for_xml_id("web.action_base_document_layout_configurator")
+        ctx = action.get('context')
+        py_ctx = json.loads(ctx) if ctx else {}
```
Comment (Julien00859) [inline odoo/addons/base/models/ir_actions_report.py:1012]:
```suggestion
        py_ctx = json.loads(action.get('context', {}))
```

File: odoo/addons/base/models/ir_actions_report.py [hunk 2]
```diff
@@ -1002,12 +1002,15 @@ def report_action(self, docids, data=None, config=True):
...
+        report_action['close_on_report_download'] = True
```
Comment (Julien00859) [inline odoo/addons/base/models/ir_actions_report.py:1013]: Default value can be None as per the function signature, `None['close_on_report_download']` is invalid. Either the function's signature is invalid, either this line is.

File: addons/web/models/base_document_layout.py [hunk 3]
```diff
@@ -40,12 +40,20 @@ def _default_report_footer(self):
...
         if 'company_name' not in address_format:
             address_format = '%(company_name)s\n' + address_format
             company_data['company_name'] = company_data['company_name'] or company.name
         return Markup(nl2br(address_format)) % company_data
 
+    def _clean_address_format(self, address_format, company_data):
+        missing_company_data = [k for k, v in company_data.items() if not v]
+        for key in missing_company_data:
+            if key in address_format:
+                address_format = ''.join(address_format.split(f'%({key})s\n'))
```
Comment (Julien00859) [inline addons/web/models/base_document_layout.py:54]: Isn't this split + join a complicated way to remove some substrings from the string?

```suggestion
                address_format = address_format.replace(f'%({key})s\n', '')
```

Are we sure of the format? It is really always `%s(...)s\n`? (real question, I'm not a PDF address expert)

**Review by Julien00859 — COMMENTED** [review]
I would like we use this opportunity to add a test case for that `_default_company_details` method. There is a test file at `/addon/web/tests/test_base_document_layout.py` already.

File: odoo/addons/base/models/ir_actions_report.py (hunk 2 above)
Comment (epictete) [reply odoo/addons/base/models/ir_actions_report.py:1013]: Indeed, it's a leftover from a previous iteration, nice catch :ok_hand: 
The signature is incorrect.

File: addons/web/models/base_document_layout.py (hunk 3 above)
Comment (epictete) [reply addons/web/models/base_document_layout.py:54]: I would say I'm "pretty sure" :wink: 
It comes from here : 
https://github.com/odoo/odoo/blob/15.0/odoo/addons/base/models/res_partner.py#L983-L985

Comment (Julien00859) [reply addons/web/models/base_document_layout.py]: https://github.com/odoo/odoo/blob/06c8765d362f81aa92f25029cb5d4e9d4069b3d3/odoo/addons/base/models/res_partner.py#L983-L985

So it is overridable and doesn't always ends with a newline (city, state_code and country_name). I don't have a good feeling about that search and replace...

Comment (epictete) [reply addons/web/models/base_document_layout.py]: I must admit I took my inspiration from what was already being done in `_default_company_details()` :
https://github.com/odoo/odoo/blob/77fce500a9742dc2bbec6b46193e3ef1546bfe85/addons/web/models/base_document_layout.py#L44-L46
We are already relying on the same `%s(...)s\n` pattern.
So I figured it was OK to follow in those footsteps :man_shrugging: 
And with the newly added test, we should know if/when `_clean_address_format()` stops working.
But maybe you have another idea/suggestion ?

**Review by oco-odoo — APPROVED** [review]
@robodoo r+

**Comment by oco-odoo** [comment]
@robodoo retry

**Comment by Xavier-Do** [comment]
Test deactivated @robodoo retry

**Comment by Xavier-Do** [comment]
@robodoo retry

**Comment by oco-odoo** [comment]
@robodoo retry

**Pull request closed without merging**